### ISSUE TITLE
ANSIBLE ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)

### ISSUE THREAD

**Issue by wannaknowmo** [ansible/ansible#14089]
hi,
i'm having an issue with ansible 2.0 when i'm trying to launch an EC2 instance. getting an ssl verification error message and i have no clue as to what to do to fix this issue :-(

i also try the callback_plugins to disable the ssl verification but i had no luck.
 cat fix.py
import ssl
if hasattr(ssl, '_create_default_https_context') and hasattr(ssl, '_create_unverified_context'):
    ssl._create_default_https_context = ssl._create_unverified_context

class CallbackModule(object):
    pass

i'm able to create/stop/terminate ec2 instances using python & boto alone. 
Voila

openssl version
OpenSSL 1.0.2d 9 Jul 2015

cat /etc/redhat-release
Red Hat Enterprise Linux Server release 6.6 (Santiago)

python --version
Python 2.7.11 :: Anaconda 2.4.1 (64-bit)

ansible --version
[WARNING]: log file at /var/log/ansible.log is not writeable, aborting
ansible 2.0.0.2
  config file = /user/wannaknowmo/ansible/cfg/ansible.cfg
  configured module search path = /user/wannaknowmo/ansible/library

ansible.cfg
[defaults]
host_key_checking = False
gathering = explicit
# gathering = false

env = DEV
transport = ssh
hostfile = /user/wannaknowmo/ansible/cfg/hosts
log_path = /var/log/ansible.log
fork=10
ANSIBLE_LIBRARY=$HOME/ansible/library
MODULE_PATH=$HOME/ansible/custom_modules

[ssh_connection]
ssh_args =""
# ssh_args = -o ControlPath=./ssh_control_%l_%h_%p_%r

pipelining = True
remote_tmp = /var/tmp/.ansible/tmp

 /user/wannaknowmo/anaconda2/bin/ansible-playbook -vvv launchEc2.yml -i ../cfg/hosts
No config file found; using defaults
1 plays in launchEc2.yml

PLAY [Provision an EC2 Instance] ***********************************************

TASK [Create a security group] *************************************************
task path: /user/wannaknowmo/ansible/playbook/launchEc2.yml:19
ESTABLISH LOCAL CONNECTION FOR USER: gfadm
localhost EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268 )" )
localhost PUT /tmp/tmpEYeaOe TO /user/wannaknowmo/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268/ec2_group
localhost EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /user/wannaknowmo/anaconda2/bin/python /user/wannaknowmo/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268/ec2_group; rm -rf "/user/wannaknowmo/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268/" > /dev/null 2>&1
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/user/wannaknowmo/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268/ec2_group", line 2632, in <module>
    main()
  File "/user/wannaknowmo/.ansible/tmp/ansible-tmp-1453591792.86-38906947896268/ec2_group", line 272, in main
    for curGroup in ec2.get_all_security_groups():
  File "/user/wannaknowmo/anaconda2/lib/python2.7/site-packages/boto/ec2/connection.py", line 2968, in get_all_security_groups
    [('item', SecurityGroup)], verb='POST')
  File "/user/wannaknowmo/anaconda2/lib/python2.7/site-packages/boto/connection.py", line 1170, in get_list
    response = self.make_request(action, params, path, verb)
  File "/user/wannaknowmo/anaconda2/lib/python2.7/site-packages/boto/connection.py", line 1116, in make_request
    return self._mexe(http_request)
  File "/user/wannaknowmo/anaconda2/lib/python2.7/site-packages/boto/connection.py", line 1030, in _mexe
    raise ex
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)

fatal: [localhost -> localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ec2_group"}, "parsed": false}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1

**Comment by chrrrles** [comment]
It can be hard to diagnose the problem without seeing the contents of the playbook.  Have you tried the `validate_certs` parameter for ec2_group ?

##### Issue Type:

Can you help us out in labelling this by telling us what kind of ticket this this?  You can say:
- Bug Report
- Feature Idea
- Feature Pull Request
- New Module Pull Request
- Bugfix Pull Request
- Documentation Report
- Docs Pull Request

##### Ansible Version:

Let us know what version of Ansible you are using.  Please supply the verbatim output from running “ansible --version”.  If you're filing a ticket on a version of Ansible which is not the latest, we'd greatly appreciate it if you could retest on the latest version first.  We don't expect you to test against the development branch most of the time, but we may ask for that if you have cycles.  Thanks!

##### Ansible Configuration:

What have you changed about your Ansible installation?  What configuration settings have you changed/added/removed?  Compare your /etc/ansible/ansible.cfg against a clean version from Github and let us know what's different.

##### Environment:

What OS are you running Ansible from and what OS are you managing?  Examples include RHEL 5/6, Centos 5/6, Ubuntu 12.04/13.10, *BSD, Solaris.  If this is a generic feature request or it doesn’t apply, just say “N/A”.  Not all tickets may be about operating system related things and we understand that.

##### Summary:

Please summarize your request in this space.  You will earn bonus points for being succinct, but please add enough detail so we can understand the request.  Thanks!

##### Steps To Reproduce:

If this is a bug ticket, please enter the steps you use to reproduce the problem in the space below.  If this is a feature request, please enter the steps you would use to use the feature.  If an example playbook is useful, please include a short reproducer inline, indented by four spaces.  If a longer one is necessary, linking to one uploaded to gist.github.com would be great.  Much appreciated!

##### Expected Results:

Please enter your expected results in this space.  When running the steps supplied above in the previous section, what did you expect to happen?  If showing example output, please indent your output by four spaces so it will render correctly in GitHub's viewer thingy.

##### Actual Results:

Please enter your actual results in this space.  When running the steps supplied above, what actually happened?  If you are showing example output, please indent your output by four spaces so it will render correctly in GitHub.  Thanks again!

**Comment by wannaknowmo** [comment]
Issue Type:
Bug Report

Ansible Version:

2.0.0.2
Ansible Configuration:

```
[defaults]
host_key_checking = False
gathering = explicit
#gathering = false
env = DEV
transport = ssh
hostfile = /user/wannaknowmo/ansible/cfg/hosts
log_path = /var/log/ansible.log
fork=10
ANSIBLE_PYTHON_INTERPRETER=/user/wannaknowmo/anaconda2/bin/python
ANSIBLE_LIBRARY=$HOME/ansible/library
MODULE_PATH=$HOME/ansible/custom_modules

[ssh_connection]
ssh_args =""
#ssh_args = -o ControlPath=./ssh_control_%l_%h_%p_%r
pipelining = True
remote_tmp = /var/tmp/.ansible/tmp
```

Environment:
Red Hat Enterprise Linux Server release 6.6 (Santiago)
Summary:

i'm having an issue with ansible 2.0 when i'm trying to launch an EC2 instance from an ansible playbook. getting an ssl verification error message and i have no clue as to what to do to fix this issue :-(

i tried a couple of of time to reinstall python 2.7.11 with boto with ansible but still no luck.
ansible & boto were reinstall with pip, python was reinstalled with anaconda bash installer.

keep in mind that i'm able to connect to aws with boto from python.
in the playbook, i use validate_certs = no but it is still doing the ssl verification

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [include_vars] ************************************************************
ok: [localhost] => {"ansible_facts": {"ec2_image": "ami-775e4f16", "ec2_instance_type": "t1.micro", "ec2_keypair": "TEST", "ec2_region": "us-east-1", "ec2_security_group": "SG_TEST", "ec2_subnet_ids": ["subnet-192.168.0.1", "subnet-192.168.1.1", "subnet-192.168.2.1"], "ec2_tag_Environment": "test", "ec2_tag_Name": "Webserver", "ec2_tag_Type": "webserver", "ec2_volume_size": 8, "validate_certs": false}, "changed": false}

TASK [provision-ec2 : Provision EC2 Box] ***************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)
fatal: [localhost -> localhost]: FAILED! => {"changed": false, "failed": true, "parsed": false}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1



 provision-ec2.yml
---
 - hosts: localhost
   connection: local
   gather_facts: true
   user: root
   pre_tasks:
    - include_vars: ec2_vars/{{type}}.yml
   roles:
    - provision-ec2

include_vars
webservers.yml
ec2_keypair: "TEST"
ec2_security_group: "SG_TEST"
ec2_instance_type: "t1.micro"
ec2_image: "ami-775e4f16"
ec2_subnet_ids: ['subnet-192.168.0.1','subnet-192.168.1.1','subnet-192.168.2.1']
ec2_region: "us-east-1"
ec2_tag_Name: "Webserver"
ec2_tag_Type: "webserver"
ec2_tag_Environment: "test"
ec2_volume_size: 8
validate_certs: no


role

---
 - name: Provision EC2 Box
   local_action:
     module: ec2
     key_name: "{{ ec2_keypair }}"
     group_id: "{{ ec2_security_group }}"
     instance_type: "{{ ec2_instance_type }}"
     image: "{{ ec2_image }}"
     vpc_subnet_id: "{{ ec2_subnet_ids|random }}"
     region: "{{ ec2_region }}"
     instance_tags: '{"Name":"{{ec2_tag_Name}}","Type":"{{ec2_tag_Type}}","Environment":"{{ec2_tag_Environment}}"}'
     assign_public_ip: no
     validate_certs: no
     wait: true
     count: 1
     volumes:
     - device_name: /dev/sda1
       device_type: gp2
       volume_size: "{{ ec2_volume_size }}"
       delete_on_termination: true
   register: ec2

 - debug: var=item
   with_items: ec2.instances

 - add_host: name={{ item.public_ip }} >
             groups=tag_Type_{{ec2_tag_Type}},tag_Environment_{{ec2_tag_Environment}}
             ec2_region={{ec2_region}}
             ec2_tag_Name={{ec2_tag_Name}}
             ec2_tag_Type={{ec2_tag_Type}}
             ec2_tag_Environment={{ec2_tag_Environment}}
             ec2_ip_address={{item.public_ip}}
   with_items: ec2.instances

 - name: Wait for the instances to boot by checking the ssh port
   wait_for: host={{item.public_ip}} port=22 delay=60 timeout=320 state=started
   with_items: ec2.instances
```

**Comment by wannaknowmo** [comment]
thanks for looking into it

**Comment by abadger** [comment]
What version of boto?

**Comment by abadger** [comment]
Note: This probably needs to move to the modules core repository: https://github.com/ansible/ansible-modules-core as it's a problem with the ec2 module.  However, there is also a chance this is in the common code shared between all the ec2 modules so I'm going to leave it here for now.

**Comment by abadger** [comment]
Note the playbook and error in the comment don't seem to match with the traceback in the initial report.  Can you post the traceback that goes along with the comment?

**Comment by abadger** [comment]
Codewise, this piece of the common code should be triggering:

```
    if HAS_BOTO3 and boto3:
[...]
        if validate_certs:
            boto_params['verify'] = validate_certs
[...]
    else:
[...]
        if validate_certs and HAS_LOOSE_VERSION and LooseVersion(boto.Version) >= LooseVersion("2.6.0"):
            boto_params['validate_certs'] = validate_certs
```

So if boto has no bugs and you're on boto 2.6.0 or greater then this should be telling boto not to validate certificates.

**Comment by abadger** [comment]
Found it -- the default for boto is to set verify/validate_certs to True.  So we have to set it explicitly to False when we get the validate_certs=False parameter.  I'll work up a fix for that.

**Comment by abadger** [comment]
That commit should fix this for you.  It's merged into the devel and stable-2.0 branches.  We'll probably create a 2.0.1rc2 next week and if so, this fix will be in it.  otherwise, the fix will be in 2.0.2.

**Comment by wannaknowmo** [comment]
I m using boto version 2.38. Thank you very much for the quick turnaround.
Will try 2.0.1.rc2 next week thanks again.
